File: Rotor.cpp
#include "Rotor.h"
#include <iostream>
#include <cstdlib>
#include <stdlib.h>
#include <ctime>
#include <time.h>
#include <string>
#include <algorithm>

Rotor::Rotor() {
    this->generateJumbledAlphabet();
}

const std::string Rotor::alphabet = "abcdefghijklmnopqrstuvwxyz";

char Rotor::getEncodedChar(char c, int rotationDegree) {
    return this->jumbledAlphabet[positiveMod((indexOf(c) - rotationDegree), 26)];
}

char Rotor::getDecodedChar(char c, int rotationDegree) {
    return this->alphabet[positiveMod((indexOfJumbled(c) + rotationDegree), 26)];
}

/* Generates jumpedAlphabet by randomly permuting elements in the 
 regular alphabet. */
void Rotor::generateJumbledAlphabet() {
    this->jumbledAlphabet = this->alphabet;
   
    int temp;
    srand(time(NULL));
    int choice;
    for(int i = 25; i > 0; i--) {
        choice = rand() % i;
        std::swap(jumbledAlphabet.at(choice), jumbledAlphabet.at(i));        
    }  
}

int Rotor::indexOf(char c) {
    for(int i = 0; i < 26; i++) {
        if(alphabet[i] == c) {
            return i; 
        }
    }
}

int Rotor::indexOfJumbled(char c) {
    for(int i = 0; i < 26; i++) {
        if(jumbledAlphabet[i] == c) {
            return i; 
        }
    }
}

int Rotor::positiveMod(int a, int b) {
    return ((a%b) + b) % b;
}

File: pythonVersion/rotor.py
import random

class Rotor:

    def swap(self, first, second):
        temp = self.jumbledAlphabet[first];
        self.jumbledAlphabet[first] = self.jumbledAlphabet[second];
        self.jumbledAlphabet[second] = temp;

    def generateJumbledAlphabet(self):
        self.jumbledAlphabet = [];

        for i in range(0, self.alphLength):
            self.jumbledAlphabet.append(self.alphabet[i]);

        for i in range(self.alphLength - 1, 0, -1):
            choice = random.choice(range(0, i));
            self.swap(i, choice);
            
    def positiveMod(self, value, mod):
        return ((value % mod) + mod) % mod;

    def getEncodedChar(self, c, rotationDegree):
        index = self.positiveMod(self.alphabet.index(c) - rotationDegree, self.alphLength); 
        return self.jumbledAlphabet[index];		

    def getDecodedChar(self, c, rotationDegree):
        index = self.positiveMod(self.jumbledAlphabet.index(c) + rotationDegree, self.alphLength); 
        return self.alphabet[index];		


    def setSecretAlphabet(self, secret):
        for i in range(0, self.alphLength):
            self.jumbledAlphabet[i] = secret[i];

    def __init__(self, alphabet):
        self.alphabet = alphabet;
        self.alphLength = len(self.alphabet);
        self.generateJumbledAlphabet();
	
